### ISSUE TITLE
Mono: ArrayType isn't an IReflectableType causes problems with GetTypeInfo

### ISSUE THREAD

**Issue by jeremymeng** [castleproject/Core#160]
Split from #155 

```
1) Test Error : Castle.DynamicProxy.Tests.ClassProxyConstructorsTestCase.ShouldGenerateTypeWithDuplicatedBaseInterfacesClassProxy
   System.InvalidCastException : Specified cast is not valid.
  at System.Reflection.IntrospectionExtensions.GetTypeInfo (System.Type type) <0x41bacc10 + 0x00070> in <filename unknown>:0
  at Castle.DynamicProxy.Generators.Emitters.SimpleAST.ConvertExpression.Emit (IMemberEmitter member, System.Reflection.Emit.ILGenerator gen) <0x4208db90 + 0x0013d> in <filename unknown>:0
  at Castle.DynamicProxy.Generators.Emitters.SimpleAST.MethodInvocationExpression.Emit (IMemberEmitter member, System.Reflection.Emit.ILGenerator gen) <0x4208d7b0 + 0x0006d> in <filename unknown>:0
  at Castle.DynamicProxy.Generators.Emitters.SimpleAST.ExpressionStatement.Emit (IMemberEmitter member, System.Reflection.Emit.ILGenerator gen) <0x4208d770 + 0x0002b> in <filename unknown>:0
  at Castle.DynamicProxy.Generators.Emitters.CodeBuilders.AbstractCodeBuilder.Generate (IMemberEmitter member, System.Reflection.Emit.ILGenerator il) <0x4208cc50 + 0x000fd> in <filename unknown>:0
  at Castle.DynamicProxy.Generators.Emitters.MethodEmitter.Generate () <0x4208d710 + 0x0004f> in <filename unknown>:0
  at Castle.DynamicProxy.Generators.Emitters.AbstractTypeEmitter.EnsureBuildersAreInAValidState () <0x4208c350 + 0x0031d> in <filename unknown>:0
  at Castle.DynamicProxy.Generators.Emitters.AbstractTypeEmitter.BuildType () <0x4208c260 + 0x00023> in <filename unknown>:0
  at Castle.DynamicProxy.Contributors.ClassProxyTargetContributor.BuildInvocationType (Castle.DynamicProxy.Generators.MetaMethod method, Castle.DynamicProxy.Generators.Emitters.ClassEmitter class, Castle.DynamicProxy.ProxyGenerationOptions options) <0x42085a80 + 0x00134> in <filename unknown>:0
  at Castle.DynamicProxy.Contributors.ClassProxyTargetContributor.GetInvocationType (Castle.DynamicProxy.Generators.MetaMethod method, Castle.DynamicProxy.Generators.Emitters.ClassEmitter class, Castle.DynamicProxy.ProxyGenerationOptions options) <0x42085a50 + 0x0001b> in <filename unknown>:0
  at Castle.DynamicProxy.Contributors.ClassProxyTargetContributor.GetMethodGenerator (Castle.DynamicProxy.Generators.MetaMethod method, Castle.DynamicProxy.Generators.Emitters.ClassEmitter class, Castle.DynamicProxy.ProxyGenerationOptions options, Castle.DynamicProxy.Contributors.OverrideMethodDelegate overrideMethod) <0x42085720 + 0x000d7> in <filename unknown>:0
  at Castle.DynamicProxy.Contributors.CompositeTypeContributor.ImplementMethod (Castle.DynamicProxy.Generators.MetaMethod method, Castle.DynamicProxy.Generators.Emitters.ClassEmitter class, Castle.DynamicProxy.ProxyGenerationOptions options, Castle.DynamicProxy.Contributors.OverrideMethodDelegate overrideMethod) <0x420855c0 + 0x0004b> in <filename unknown>:0
  at Castle.DynamicProxy.Contributors.CompositeTypeContributor.Generate (Castle.DynamicProxy.Generators.Emitters.ClassEmitter class, Castle.DynamicProxy.ProxyGenerationOptions options) <0x42085130 + 0x00133> in <filename unknown>:0
  at Castle.DynamicProxy.Generators.ClassProxyGenerator.GenerateType (System.String name, System.Type[] interfaces, INamingScope namingScope) <0x4206e6d0 + 0x002d8> in <filename unknown>:0
  at Castle.DynamicProxy.Generators.ClassProxyGenerator+<GenerateCode>c__AnonStorey0.<>m__0 (System.String n, INamingScope s) <0x4206e690 + 0x0002f> in <filename unknown>:0
  at Castle.DynamicProxy.Generators.BaseProxyGenerator.ObtainProxyType (Castle.DynamicProxy.Generators.CacheKey cacheKey, System.Func`3 factory) <0x4206dbf0 + 0x003cf> in <filename unknown>:0
  at Castle.DynamicProxy.Generators.ClassProxyGenerator.GenerateCode (System.Type[] interfaces, Castle.DynamicProxy.ProxyGenerationOptions options) <0x4206cbe0 + 0x00257> in <filename unknown>:0
  at Castle.DynamicProxy.DefaultProxyBuilder.CreateClassProxyType (System.Type classToProxy, System.Type[] additionalInterfacesToProxy, Castle.DynamicProxy.ProxyGenerationOptions options) <0x4206c470 + 0x000cb> in <filename unknown>:0
  at Castle.DynamicProxy.ProxyGenerator.CreateClassProxyType (System.Type classToProxy, System.Type[] additionalInterfacesToProxy, Castle.DynamicProxy.ProxyGenerationOptions options) <0x4206c420 + 0x0003f> in <filename unknown>:0
  at Castle.DynamicProxy.ProxyGenerator.CreateClassProxy (System.Type classToProxy, System.Type[] additionalInterfacesToProxy, Castle.DynamicProxy.ProxyGenerationOptions options, System.Object[] constructorArguments, Castle.DynamicProxy.IInterceptor[] interceptors) <0x4206c030 + 0x000bb> in <filename unknown>:0
  at Castle.DynamicProxy.Tests.ClassProxyConstructorsTestCase.ShouldGenerateTypeWithDuplicatedBaseInterfacesClassProxy () <0x420c7d70 + 0x000e3> in <filename unknown>:0
  at (wrapper managed-to-native) System.Reflection.MonoMethod:InternalInvoke (System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) <0x41bad940 + 0x000b7> in <filename unknown>:0
```

The other two failures are similar, also failing in `ConvertExpression`

```
2) Test Error : Castle.DynamicProxy.Tests.GenericClassProxyTestCase.TypeWithGenericMethodHavingArgumentBeingGenericArrayOfT
   System.InvalidCastException : Specified cast is not valid.
  at System.Reflection.IntrospectionExtensions.GetTypeInfo (System.Type type) <0x41e02b10 + 0x00070> in <filename unknown>:0
  at Castle.DynamicProxy.Generators.Emitters.SimpleAST.ConvertExpression.Emit (IMemberEmitter member, System.Reflection.Emit.ILGenerator gen) <0x417648f0 + 0x0013d> in <filename unknown>:0
  at Castle.DynamicProxy.Generators.Emitters.SimpleAST.MethodInvocationExpression.Emit (IMemberEmitter member, System.Reflection.Emit.ILGenerator gen) <0x41764510 + 0x0006d> in <filename unknown>:0
  at Castle.DynamicProxy.Generators.Emitters.SimpleAST.AssignStatement.Emit (IMemberEmitter member, System.Reflection.Emit.ILGenerator gen) <0x4176e2d0 + 0x00044> in <filename unknown>:0
  at Castle.DynamicProxy.Generators.Emitters.CodeBuilders.AbstractCodeBuilder.Generate (IMemberEmitter member, System.Reflection.Emit.ILGenerator il) <0x417639b0 + 0x000fd> in <filename unknown>:0
  at Castle.DynamicProxy.Generators.Emitters.MethodEmitter.Generate () <0x41764470 + 0x0004f> in <filename unknown>:0
```
and
```
3) Test Error : Castle.DynamicProxy.Tests.GenericMethodsProxyTestCase.GenericMethod_WithArrayOfGenericOfGenericArgument
   System.InvalidCastException : Specified cast is not valid.
  at System.Reflection.IntrospectionExtensions.GetTypeInfo (System.Type type) <0x41e02b10 + 0x00070> in <filename unknown>:0
  at Castle.DynamicProxy.Generators.Emitters.SimpleAST.ConvertExpression.Emit (IMemberEmitter member, System.Reflection.Emit.ILGenerator gen) <0x417648f0 + 0x0013d> in <filename unknown>:0
```

Some progress.  All three tests failed on calling `target.GetTypeInfo()` when `target` is a `System.Reflection.Emit.ArrayType`, which is not an `IReflectableType`.

The parameter type is from https://github.com/castleproject/Core/blob/430d0282e9eab828ec7de8f0e37479ce630d98f9/src/Castle.Core/DynamicProxy/Generators/InvocationTypeGenerator.cs#L135. 
Step into the method, the array element type is assigned at https://github.com/castleproject/Core/blob/430d0282e9eab828ec7de8f0e37479ce630d98f9/src/Castle.Core/DynamicProxy/Internal/TypeUtil.cs#L119, which in turn calls the recursive method, and returned at https://github.com/castleproject/Core/blob/430d0282e9eab828ec7de8f0e37479ce630d98f9/src/Castle.Core/DynamicProxy/Internal/TypeUtil.cs#L108.  

When debugging on Windows,  the line `return parameter.GetGenericTypeDefinition().MakeGenericType(arguments);` returns

```
+		elementType	{Name = "List`1" FullName = null}	System.Type {System.Reflection.Emit.TypeBuilderInstantiation}
```

while Console.WriteLine debugging on Mono says

```
System.Reflection.MonoGenericClass
```

Following this, on Mono, `elementType.MakeArrayType()` gives a type of  `System.Reflection.Emit.ArrayType` that is not `IReflectableType`.   

On windows it is
```
+		elementType.MakeArrayType()	{Name = "List`1[]" FullName = null}	System.Type {System.Reflection.Emit.SymbolType}
```

I haven't gone deeper in Mono code yet.  It seems a implementation difference.

**Comment by jeremymeng** [comment]
I came up with a simpler test file.  On Windows it outputs `System.RuntimeType`, on Mono 4.2.2 it outputs `System.MonoType`.  Logged [a bug](https://bugzilla.xamarin.com/show_bug.cgi?id=39700) against mono and see what they say:

**Comment by jonorossi** [comment]
@jeremymeng I agree with Zoltan Varga (in [Bug 39700](https://bugzilla.xamarin.com/show_bug.cgi?id=39700)), `MonoType` is an implementation detail that isn't a defect, they just have another type in their implementation that adds to `RuntimeType`. We need to provide the Mono guys with a actual test case to include the use of `GetTypeInfo` that breaks because it requires an `IReflectableType`.

Starting off I noticed the exception message is different on different Mono versions, 4.0.2 reports [`System.InvalidCastException: Cannot cast from source type to destination type.`](https://github.com/mono/mono/blob/mono-4.0.2.5/mcs/class/corlib/System/InvalidCastException.cs#L48) while 4.2.2 reports [`System.InvalidCastException : Specified cast is not valid.`](https://github.com/mono/referencesource/blob/mono/mscorlib/mscorlib.txt#L236) because the new code uses the MS resource strings.

Looking at the Mono code `System.Reflection.MonoGenericClass` is the name they gave to the internal class which represents the same concept as `System.Reflection.Emit.TypeBuilderInstantiation` in .NET Framework.

Now looking into why `ArrayType` isn't an `IReflectableType`.

On .NET Framework:
* `Type`, is a `MemberInfo`
* `TypeInfo`, is a `Type` and an `IReflectableType`
* `RuntimeType`, is a `TypeInfo`
* `SymbolType`, is a `TypeInfo`

On Mono:
* [`Type`](https://github.com/mono/mono/blob/master/mcs/class/corlib/ReferenceSources/Type.cs#L35)[[2](https://github.com/mono/referencesource/blob/mono/mscorlib/system/type.cs#L41)], is a [`MemberInfo`](https://github.com/mono/referencesource/blob/mono/mscorlib/system/reflection/memberinfo.cs#L28)
* [`TypeInfo`](https://github.com/mono/referencesource/blob/mono/mscorlib/system/reflection/typeinfo.cs#L29), is a `Type` and an `IReflectableType`
* [`MonoType`](https://github.com/mono/mono/blob/master/mcs/class/corlib/System/MonoType.cs#L63), is a [`RuntimeType`](https://github.com/mono/referencesource/blob/mono/mscorlib/system/rttype.cs#L94-L95), is a `TypeInfo`
* [`ArrayType`](https://github.com/mono/mono/blob/b7a308f660de8174b64697a422abfc7315d07b8c/mcs/class/corlib/System.Reflection.Emit/DerivedTypes.cs#L329), is a [`DerivedType`](https://github.com/mono/mono/blob/b7a308f660de8174b64697a422abfc7315d07b8c/mcs/class/corlib/System.Reflection.Emit/DerivedTypes.cs#L49), is a `Type`
* `ByRefType`, is a `DerivedType`
* `PointerType`, is a `DerivedType`

It appears as though `TypeInfo` wasn't pushed into the hierarchy of `DerivedType` when it was added to `RuntimeType` in Mono, which results in `GetTypeInfo` falling over because it assumes all publicly accessible `System.Type`s implement `IReflectableType`.

**Comment by jonorossi** [comment]
Here is a unit test:

```csharp
using System;
using System.Reflection;
using System.Reflection.Emit;

namespace ConsoleApplication3
{
	public class Program
	{
		public static void Main(string[] args)
		{
			AssemblyBuilder assemblyBuilder = AppDomain.CurrentDomain.DefineDynamicAssembly(new AssemblyName("TestDynAssembly"), AssemblyBuilderAccess.Run);
			ModuleBuilder moduleBuilder = assemblyBuilder.DefineDynamicModule("TestDynModule");
			TypeBuilder typeBuilder = moduleBuilder.DefineType("TestDynType", TypeAttributes.Public);
			GenericTypeParameterBuilder[] typeParamBuilders = typeBuilder.DefineGenericParameters(new[] { "T" });

			Type arrayType = typeParamBuilders[0].MakeArrayType();
			Type twoDimArrayType = typeParamBuilders[0].MakeArrayType(2);
			Type byRefType = typeParamBuilders[0].MakeByRefType();
			Type pointerType = typeParamBuilders[0].MakePointerType();

			Type[] parameterTypes = { arrayType, twoDimArrayType, byRefType, pointerType };
			foreach (Type t in parameterTypes)
			{
				Console.WriteLine(t.GetType().ToString());
				try
				{
					Console.WriteLine(t.GetTypeInfo().ToString());
				}
				catch (Exception ex)
				{
					Console.WriteLine(ex.Message);
				}
			}
		}
	}
}
```

.NET Framework output:
```
System.Reflection.Emit.SymbolType
T[]
System.Reflection.Emit.SymbolType
T[,]
System.Reflection.Emit.SymbolType
T&
System.Reflection.Emit.SymbolType
T*
```

Mono 4.0.2 output:
```
System.Reflection.Emit.ArrayType
Cannot cast from source type to destination type.
System.Reflection.Emit.ArrayType
Cannot cast from source type to destination type.
System.Reflection.Emit.ByRefType
Cannot cast from source type to destination type.
System.Reflection.Emit.PointerType
Cannot cast from source type to destination type.
```

**Comment by jonorossi** [comment]
The title and description of Mono [Bug 39700](https://bugzilla.xamarin.com/show_bug.cgi?id=39700) doesn't actually describe the Mono defect. Let me know if you are happy with my findings and we should log a new Mono issue.

**Comment by jeremymeng** [comment]
@jonorossi your example look good.  I had thought that `elementType.MakeArrayType()` shouldn't have returned `ArrayType` thus went into that direction.  I will close the Mono bug 39700.

**Comment by jonorossi** [comment]
I've logged [Bug 39911](https://bugzilla.xamarin.com/show_bug.cgi?id=39911) with Mono.

**Comment by jonorossi** [comment]
The defect has been fixed in Mono on the master branch, we'll retest this when the next Mono release comes out.

https://github.com/mono/mono/commit/17193833a59194099f76d110a4500fae07a5b1cb

**Comment by jonorossi** [comment]
This fix still hasn't shipped, Mono 4.4.1.0 that came out last month didn't include it.

**Comment by jonorossi** [comment]
This fix is included in Mono 4.6.0 that was released last month, will upgrade our CI server and retest.

**Comment by jonorossi** [comment]
I've upgraded our Linux build server to Mono 4.6.1, and fixed a build script problem with the new XBuild and some warnings in b20896812651887294184c09dee8d824e6179c46. These unit tests are now passing.
